--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -37,6 +37,17 @@
   ADD_DEFINITIONS("-D_GNU_SOURCE")
 endif("${CYGWIN}")
 
+get_property(LIB64 GLOBAL PROPERTY FIND_LIBRARY_USE_LIB64_PATHS)
+
+if ("${LIB64}" STREQUAL "TRUE")
+    set(LIBSUFFIX 64)
+else()
+    set(LIBSUFFIX "")
+endif()
+
+set(INSTALL_LIB_DIR     lib${LIBSUFFIX} CACHE PATH "Installation directory for libraries")
+mark_as_advanced(INSTALL_LIB_DIR)
+
 SET(CMAKE_EXE_LINKER_FLAGS_PROFILE "${CMAKE_EXE_LINKER_FLAGS_PROFILE} -pg")
 SET(CMAKE_SHARED_LINKER_FLAGS_PROFILE "${CMAKE_SHARED_LINKER_FLAGS_PROFILE} -pg")
 
--- a/src/librexgen/lua/CMakeLists.txt
+++ b/src/librexgen/lua/CMakeLists.txt
@@ -35,8 +35,8 @@
 
 install(TARGETS ${librexgen_lua}
 		RUNTIME DESTINATION "bin"
-		LIBRARY DESTINATION "lib"
-		ARCHIVE DESTINATION "lib"
+		LIBRARY DESTINATION "${INSTALL_LIB_DIR}"
+		ARCHIVE DESTINATION "${INSTALL_LIB_DIR}"
 		)
 
 SET(INSTALL_INCLUDE_DIR "include/${librexgen}")
--- a/src/librexgen/CMakeLists.txt
+++ b/src/librexgen/CMakeLists.txt
@@ -98,8 +98,8 @@
 
 install(TARGETS ${librexgen}
 		RUNTIME DESTINATION "bin"
-		LIBRARY DESTINATION "lib"
-		ARCHIVE DESTINATION "lib"
+		LIBRARY DESTINATION "${INSTALL_LIB_DIR}"
+		ARCHIVE DESTINATION "${INSTALL_LIB_DIR}"
 		)
 
 SET(INSTALL_INCLUDE_DIR "include/${librexgen}")
--- a/src/rexgen/CMakeLists.txt
+++ b/src/rexgen/CMakeLists.txt
@@ -23,5 +23,6 @@
 		OUTPUT rexgen.1.gz
 		COMMAND gzip "${CMAKE_CURRENT_SOURCE_DIR}/rexgen.1" -c >rexgen.1.gz)
 	add_custom_target(man ALL DEPENDS "${CMAKE_CURRENT_BINARY_DIR}/rexgen.1.gz")
-	install(FILES "${CMAKE_CURRENT_BINARY_DIR}/rexgen.1.gz" DESTINATION "man/man1")
+	install(FILES "${CMAKE_CURRENT_BINARY_DIR}/rexgen.1.gz" DESTINATION
+		"share/man/man1")
 endif(MSVC)
