--- a/kernel/nv-linux.h
+++ b/kernel/nv-linux.h
@@ -2093,11 +2093,39 @@
  * write and force parameters were replaced with gup_flags by - 
  * 2016 Oct 12: 768ae309a96103ed02eb1e111e838c87854d8b51
  *
+ * A 7-argument version of get_user_pages was introduced into linux-4.4.y by
+ * commit 8e50b8b07f462ab4b91bc1491b1c91bd75e4ad40 which cherry-picked the
+ * replacement of the write and force parameters with gup_flags
+ *
+ *
  */
 
 #if defined(NV_GET_USER_PAGES_HAS_TASK_STRUCT)
-    #define NV_GET_USER_PAGES(start, nr_pages, write, force, pages, vmas) \
-        get_user_pages(current, current->mm, start, nr_pages, write, force, pages, vmas)
+    #if defined(NV_GET_USER_PAGES_HAS_WRITE_AND_FORCE_ARGS)
+        #define NV_GET_USER_PAGES(start, nr_pages, write, force, pages, vmas) \
+            get_user_pages(current, current->mm, start, nr_pages, write, force, pages, vmas)
+    #else
+        #include <linux/mm.h>
+        #include <linux/sched.h>
+
+        static inline long NV_GET_USER_PAGES(unsigned long start,
+                                             unsigned long nr_pages,
+                                             int write,
+                                             int force,
+                                             struct page **pages,
+                                             struct vm_area_struct **vmas)
+        {
+            unsigned int flags = 0;
+
+            if (write)
+                flags |= FOLL_WRITE;
+            if (force)
+                flags |= FOLL_FORCE;
+
+            return get_user_pages(current, current->mm, start, nr_pages, flags,
+                                  pages, vmas);
+        }
+    #endif
 #else
     #if defined(NV_GET_USER_PAGES_HAS_WRITE_AND_FORCE_ARGS)
         #define NV_GET_USER_PAGES get_user_pages
@@ -2179,7 +2207,31 @@
         }
     #endif
 #else
-    #define NV_GET_USER_PAGES_REMOTE    NV_GET_USER_PAGES
+    #if defined(NV_GET_USER_PAGES_HAS_WRITE_AND_FORCE_ARGS)
+        #define NV_GET_USER_PAGES_REMOTE    get_user_pages
+    #else
+        #include <linux/mm.h>
+        #include <linux/sched.h>
+
+        static inline long NV_GET_USER_PAGES_REMOTE(struct task_struct *tsk,
+                                                    struct mm_struct *mm,
+                                                    unsigned long start,
+                                                    unsigned long nr_pages,
+                                                    int write,
+                                                    int force,
+                                                    struct page **pages,
+                                                    struct vm_area_struct **vmas)
+        {
+            unsigned int flags = 0;
+
+            if (write)
+                flags |= FOLL_WRITE;
+            if (force)
+                flags |= FOLL_FORCE;
+
+            return get_user_pages(tsk, mm, start, nr_pages, flags, pages, vmas);
+        }
+    #endif
 #endif
 
 #endif  /* _NV_LINUX_H_ */
--- a/kernel/conftest.sh
+++ b/kernel/conftest.sh
@@ -1796,21 +1796,28 @@
             # Conftest for get_user_pages()
             #
             # Use long type for get_user_pages and unsigned long for nr_pages
-            # 2013 Feb 22: 28a35716d317980ae9bc2ff2f84c33a3cda9e884
+            # by commit 28a35716d317 ("mm: use long type for page counts
+            # in mm_populate() and get_user_pages()") in v3.9 (2013-02-22)
             #
-            # Removed struct task_struct *tsk & struct mm_struct *mm from get_user_pages.
-            # 2016 Feb 12: cde70140fed8429acf7a14e2e2cbd3e329036653
+            # Removed struct task_struct *tsk & struct mm_struct *mm from
+            # get_user_pages by commit cde70140fed8 ("mm/gup: Overload
+            # get_user_pages() functions") in v4.6 (2016-02-12)
+            #
+            # Replaced get_user_pages6 with get_user_pages by commit
+            # c12d2da56d0e ("mm/gup: Remove the macro overload API migration
+            # helpers from the get_user*() APIs") in v4.6 (2016-04-04)
+            #
+            # Replaced write and force parameters with gup_flags by
+            # commit 768ae309a961 ("mm: replace get_user_pages() write/force
+            # parameters with gup_flags") in v4.9 (2016-10-13)
             #
-            # Replaced get_user_pages6 with get_user_pages.
-            # 2016 April 4: c12d2da56d0e07d230968ee2305aaa86b93a6832
-            #
-            # Replaced write and force parameters with gup_flags.
-            # 2016 Oct 12: 768ae309a96103ed02eb1e111e838c87854d8b51
+            # linux-4.4.168 cherry-picked commit 768ae309a961 without
+            # c12d2da56d0e which is covered in Conftest #3.
             #
             # Conftest #1: Check if get_user_pages accepts 6 arguments.
             # Return if true.
             # Fall through to conftest #2 on failure.
-
+            #
             echo "$CONFTEST_PREAMBLE
             #include <linux/mm.h>
             long get_user_pages(unsigned long start,
@@ -1854,7 +1861,36 @@
                 rm -f conftest$$.o
                 return
             fi
-            
+
+            # Conftest #3: Check if get_user_pages has gup_flags instead of
+            # write and force parameters AND that gup has task_struct and
+            # mm_struct as its first arguments.
+            # Return if available.
+            # Fall through to default case if absent.
+
+            echo "$CONFTEST_PREAMBLE
+            #include <linux/mm.h>
+            long get_user_pages(struct task_struct *tsk,
+                                struct mm_struct *mm,
+                                unsigned long start,
+                                unsigned long nr_pages,
+                                unsigned int gup_flags,
+                                struct page **pages,
+                                struct vm_area_struct **vmas) {
+                return 0;
+            }" > conftest$$.c
+
+            $CC $CFLAGS -c conftest$$.c > /dev/null 2>&1
+            rm -f conftest$$.c
+
+            if [ -f conftest$$.o ]; then
+                echo "#undef NV_GET_USER_PAGES_HAS_WRITE_AND_FORCE_ARGS" | append_conftest "functions"
+                echo "#define NV_GET_USER_PAGES_HAS_TASK_STRUCT" | append_conftest "functions"
+                rm -f conftest$$.o
+                return
+            fi
+
+ 
             echo "#define NV_GET_USER_PAGES_HAS_WRITE_AND_FORCE_ARGS" | append_conftest "functions"
             echo "#define NV_GET_USER_PAGES_HAS_TASK_STRUCT" | append_conftest "functions"
 
@@ -1866,21 +1902,23 @@
             # Determine if the function get_user_pages_remote() is
             # present and has write/force parameters.
             #
-            # get_user_pages_remote() was added by:
-            #   2016 Feb 12: 1e9877902dc7e11d2be038371c6fbf2dfcd469d7
+            # get_user_pages_remote() was added by commit 1e9877902dc7
+            # ("mm/gup: Introduce get_user_pages_remote()") in v4.6 (2016-02-12)
             #
             # get_user_pages[_remote]() write/force parameters
-            # replaced with gup_flags:
-            #   2016 Oct 12: 768ae309a96103ed02eb1e111e838c87854d8b51
-            #   2016 Oct 12: 9beae1ea89305a9667ceaab6d0bf46a045ad71e7
-            #
-            # get_user_pages_remote() added 'locked' parameter
-            #   2016 Dec 14:5b56d49fc31dbb0487e14ead790fc81ca9fb2c99
+            # replaced with gup_flags by commits 768ae309a961 ("mm: replace
+            # get_user_pages() write/force parameters with gup_flags") and
+            # commit 9beae1ea8930 ("mm: replace get_user_pages_remote()
+            # write/force parameters with gup_flags") in v4.9 (2016-10-13)
+            #
+            # get_user_pages_remote() added 'locked' parameter by
+            # commit 5b56d49fc31d ("mm: add locked parameter to
+            # get_user_pages_remote()") in v4.10 (2016-12-14)
             #
             # conftest #1: check if get_user_pages_remote() is available
             # return if not available.
             # Fall through to conftest #2 if it is present
-
+            #
             echo "$CONFTEST_PREAMBLE
             #include <linux/mm.h>
             int conftest_get_user_pages_remote(void) {
