--- wiringPi/Makefile	2017-04-18 22:50:38.951022128 +0100
+++ wiringPi/Makefile	2017-04-18 22:51:19.747265645 +0100
@@ -104,8 +104,7 @@
 	$Q echo "[Install Dynamic Lib]"
 	$Q install -m 0755 -d						$(DESTDIR)$(PREFIX)/lib
 	$Q install -m 0755 libwiringPi.so.$(VERSION)			$(DESTDIR)$(PREFIX)/lib/libwiringPi.so.$(VERSION)
-	$Q ln -sf $(DESTDIR)$(PREFIX)/lib/libwiringPi.so.$(VERSION)	$(DESTDIR)/lib/libwiringPi.so
-	$Q $(LDCONFIG)
+	$Q ln -sf --relative $(DESTDIR)$(PREFIX)/lib/libwiringPi.so.$(VERSION)	$(DESTDIR)/lib/libwiringPi.so
 
 .PHONY:	install-static
 install-static:	$(STATIC)
