--- deadbeef-0.5.1/work/deadbeef-0.5.1/plugins/gtkui/gtkui.c.old	2011-05-22 17:58:53.000000000 +0400
+++ deadbeef-0.5.1/work/deadbeef-0.5.1/plugins/gtkui/gtkui.c	2011-11-20 18:56:06.000000000 +0400
@@ -545,7 +545,7 @@
 
     if (!gtk_icon_theme_has_icon(theme, icon_name)) {
         char iconpath[1024];
-        snprintf (iconpath, sizeof (iconpath), "%s/deadbeef.png", deadbeef->get_prefix ());
+        snprintf (iconpath, sizeof (iconpath), "%s/share/icons/hicolor/16x16/apps/deadbeef.png", deadbeef->get_prefix ());
         trayicon = gtk_status_icon_new_from_file(iconpath);
     }
     else {
@@ -1042,7 +1042,7 @@
     }
     else {
         char iconpath[1024];
-        snprintf (iconpath, sizeof (iconpath), "%s/deadbeef.png", deadbeef->get_prefix ());
+        snprintf (iconpath, sizeof (iconpath), "%s/share/icons/hicolor/16x16/apps/deadbeef.png", deadbeef->get_prefix ());
         gtk_window_set_icon_from_file (GTK_WINDOW (mainwin), iconpath, NULL);
     }
 
