From b14550cb539cc7f8f7497a10d31c46f22ab65b25 Mon Sep 17 00:00:00 2001
From: Hannes Schueller <hannes@yllr.net>
Date: Fri, 26 Dec 2014 16:22:21 +0100
Subject: [PATCH 15/20] contributors list

---
 LICENSE | 8 +++++---
 main.c  | 8 +++++---
 2 files changed, 10 insertions(+), 6 deletions(-)

diff --git a/LICENSE b/LICENSE
index 0c60893..797fd70 100644
--- a/LICENSE
+++ b/LICENSE
@@ -1,11 +1,13 @@
 Copyright (c) 2009 Leon Winter
-Copyright (c) 2009-2012 Hannes Schueller
+Copyright (c) 2009-2014 Hannes Schueller
 Copyright (c) 2009-2010 Matto Fransen
 Copyright (c) 2010-2011 Hans-Peter Deifel
 Copyright (c) 2010-2011 Thomas Adam
 Copyright (c) 2011 Albert Kim
-Copyright (c) 2011 Daniel Carl
-Copyright (c) 2012 Matthew Carter
+Copyright (c) 2011-2014 Daniel Carl
+Copyright (c) 2012-2014 Matthew Carter
+Copyright (c) 2014 by Morgan Howe
+Copyright (c) 2014 by Jan Niemeyer
 
 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
diff --git a/main.c b/main.c
index ca0bebe..2cb5904 100644
--- a/main.c
+++ b/main.c
@@ -1,12 +1,14 @@
 /*
     (c) 2009 by Leon Winter
-    (c) 2009-2013 by Hannes Schueller
+    (c) 2009-2014 by Hannes Schueller
     (c) 2009-2010 by Matto Fransen
     (c) 2010-2011 by Hans-Peter Deifel
     (c) 2010-2011 by Thomas Adam
     (c) 2011 by Albert Kim
-    (c) 2011-2013 by Daniel Carl
-    (c) 2012 by Matthew Carter
+    (c) 2011-2014 by Daniel Carl
+    (c) 2012-2014 by Matthew Carter
+    (c) 2014 by Morgan Howe
+    (c) 2014 by Jan Niemeyer
     see LICENSE file
 */
 
-- 
2.1.4

