From 809853f87197969a03e66ce0e202bca0f58acb6c Mon Sep 17 00:00:00 2001
From: Hannes Schueller <hannes@yllr.net>
Date: Mon, 16 Feb 2015 08:36:09 +0000
Subject: [PATCH 16/20] bugfix: re-enable downloading of files by setting
 correct HTTP status

---
 main.c | 10 ++++------
 main.h |  3 ++-
 2 files changed, 6 insertions(+), 7 deletions(-)

diff --git a/main.c b/main.c
index 2cb5904..0a24d92 100644
--- a/main.c
+++ b/main.c
@@ -1,6 +1,6 @@
 /*
     (c) 2009 by Leon Winter
-    (c) 2009-2014 by Hannes Schueller
+    (c) 2009-2015 by Hannes Schueller
     (c) 2009-2010 by Matto Fransen
     (c) 2010-2011 by Hans-Peter Deifel
     (c) 2010-2011 by Thomas Adam
@@ -239,11 +239,8 @@ webview_new_window_cb(WebKitWebView *webview, WebKitWebFrame *frame, WebKitNetwo
 gboolean
 webview_mimetype_cb(WebKitWebView *webview, WebKitWebFrame *frame, WebKitNetworkRequest *request,
                         char *mime_type, WebKitWebPolicyDecision *decision, gpointer user_data) {
-    SoupMessage *msg = webkit_network_request_get_message(request);
-    guint http_status = msg->status_code;
-
     if (webkit_web_view_can_show_mime_type(webview, mime_type) == FALSE) {
-        if (SOUP_STATUS_IS_SUCCESSFUL(http_status)) {
+        if (SOUP_STATUS_IS_SUCCESSFUL(client.net.http_status)) {
             webkit_web_policy_decision_download(decision);
             return TRUE;
         }
@@ -3020,7 +3017,8 @@ handle_response_headers(SoupMessage *soup_msg, gpointer unused)
 	GSList *resp_cookie = NULL, *cookie_list;
 	SoupCookie *cookie;
 	SoupURI *uri = soup_message_get_uri(soup_msg);
-
+	
+	client.net.http_status = soup_msg->status_code;
 	if (CookiePolicy != SOUP_COOKIE_JAR_ACCEPT_NEVER) {
 		cookie_list = soup_cookies_from_response(soup_msg);
 		for(resp_cookie = cookie_list; resp_cookie; resp_cookie = g_slist_next(resp_cookie))
diff --git a/main.h b/main.h
index 3f30225..4636257 100644
--- a/main.h
+++ b/main.h
@@ -1,6 +1,6 @@
 /*
     (c) 2009 by Leon Winter
-    (c) 2009, 2010 by Hannes Schueller
+    (c) 2009-2015 by Hannes Schueller
     (c) 2009, 2010 by Matto Fransen
     (c) 2010 by Hans-Peter Deifel
     (c) 2010 by Thomas Adam
@@ -59,6 +59,7 @@ typedef struct {
     SoupCookieJar   *session_cookie_jar;
     SoupCookieJar   *file_cookie_jar;
     char            *cookie_store;
+    guint           http_status;
 } Network;
 
 typedef struct {
-- 
2.1.4

