--- a/src/Cutter.pro	2019-04-23 13:05:32.106666667 -0000
+++ b/src/Cutter.pro	2019-04-23 13:07:01.770000001 -0000
@@ -120,10 +120,13 @@
             LIBS += -F$$PYTHON_FRAMEWORK_DIR -framework Python
             DEFINES += MACOS_PYTHON_FRAMEWORK_BUNDLED
         } else {
-            !packagesExist(python3) {
-                error("ERROR: Python 3 could not be found. Make sure it is available to pkg-config.")
+		system(type python3-config) {
+			LIBS += $$system(python3-config --libs)
+			TMP = $$system(python3-config --includes)
+			INCLUDEPATH += $$replace(TMP, "-I", "")
+		} else {
+			error("ERROR: Python 3 could not be found. Make sure it is available to python3-config.")
             }
-            PKGCONFIG += python3
         }
     }
 
