--- a/config.mk
+++ b/config.mk
@@ -1,8 +1,8 @@
 CXX ?= clang++
 CC ?= clang
-LDLIBS = -lm -lstdc++
-CFLAGS = -MD -O0 -ggdb -Wall -std=c99 -I/usr//include
-CXXFLAGS = -MD -O0 -ggdb -Wall -std=c++11 -I/usr//include
+override LDLIBS += -lm -lstdc++
+override CFLAGS += -Wall -std=c99 -I/usr/include
+override CXXFLAGS += -Wall -std=c++11 -I/usr/include
 PKG_CONFIG ?= pkg-config
 DESTDIR ?=
 PREFIX ?= /usr/
