--- a/bgpd/bgp_route.c
+++ b/bgpd/bgp_route.c
@@ -6348,13 +6348,6 @@
 						BUFSIZ));
 			destination = ntohl(p->u.prefix4.s_addr);
 
-			if ((IN_CLASSC(destination) && p->prefixlen == 24)
-			    || (IN_CLASSB(destination) && p->prefixlen == 16)
-			    || (IN_CLASSA(destination) && p->prefixlen == 8)
-			    || p->u.prefix4.s_addr == 0) {
-				/* When mask is natural,
-				   mask is not displayed. */
-			} else
 				len += vty_out(vty, "/%d", p->prefixlen);
 		} else {
 			json_object_string_add(json, "prefix",
@@ -11260,12 +11253,6 @@
 				inet_ntop(p->family, &p->u.prefix, buf,
 					  SU_ADDRSTRLEN));
 
-			if ((IN_CLASSC(destination) && p->prefixlen == 24)
-			    || (IN_CLASSB(destination) && p->prefixlen == 16)
-			    || (IN_CLASSA(destination) && p->prefixlen == 8)
-			    || p->u.prefix4.s_addr == 0) {
-				/* Natural mask is not display. */
-			} else
 				vty_out(vty, " mask %s", inet_ntoa(netmask));
 		} else {
 			vty_out(vty, "  network %s/%d",
