Index: vimprobable2-1.4.2/config.h
===================================================================
--- vimprobable2-1.4.2.orig/config.h
+++ vimprobable2-1.4.2/config.h
@@ -121,8 +121,11 @@ static Searchengine searchengines[] = {
     { "s",          "http://startpage.com/do/metasearch.pl?query=%s" },
     { "w",          "https://secure.wikimedia.org/wikipedia/en/w/index.php?title=Special%%3ASearch&search=%s&go=Go" },
     { "wd",         "https://secure.wikimedia.org/wikipedia/de/w/index.php?title=Special%%3ASearch&search=%s&go=Go" },
-    { "d",          "https://duckduckgo.com/?q=%s&t=vimprobable" },
-    { "dd",         "https://duckduckgo.com/html/?q=%s&t=vimprobable" },
+    { "d",          "https://duckduckgo.com/html/?q=%s&t=vimprobable" },
+    { "dd",         "https://duckduckgo.com/?q=%s&t=vimprobable" },
+    { "m",	    "http://manpages.ubuntu.com/cgi-bin/search.py?q=%s" },
+    { "ii",	    "http://www.imdb.com/find?q=%s" },
+    { "l",	    "http://lxr.free-electrons.com/ident?i=%s" },
 };
 
 static char defaultsearch[MAX_SETTING_SIZE] = "d";
