--- a/xpdf/PDFCore.cc
+++ b/xpdf/PDFCore.cc
@@ -1677,9 +1677,6 @@
   int x0, y0, x1, y1, t;
   GString *s;
 
-  if (!doc->okToCopy()) {
-    return NULL;
-  }
   if ((page = findPage(pg))) {
     cvtUserToDev(pg, xMin, yMin, &x0, &y0);
     cvtUserToDev(pg, xMax, yMax, &x1, &y1);
--- a/xpdf/pdfimages.cc
+++ b/xpdf/pdfimages.cc
@@ -118,14 +118,6 @@
     goto err1;
   }
 
-  // check for copy permission
-  if (!doc->okToCopy()) {
-    error(errNotAllowed, -1,
-	  "Copying of images from this document is not allowed.");
-    exitCode = 3;
-    goto err1;
-  }
-
   // get page range
   if (firstPage < 1)
     firstPage = 1;
--- a/xpdf/pdfinfo.cc
+++ b/xpdf/pdfinfo.cc
@@ -208,7 +208,6 @@
   if (doc->isEncrypted()) {
     printf("yes (print:%s copy:%s change:%s addNotes:%s)\n",
 	   doc->okToPrint(gTrue) ? "yes" : "no",
-	   doc->okToCopy(gTrue) ? "yes" : "no",
 	   doc->okToChange(gTrue) ? "yes" : "no",
 	   doc->okToAddNotes(gTrue) ? "yes" : "no");
   } else {
--- a/xpdf/pdftohtml.cc
+++ b/xpdf/pdftohtml.cc
@@ -130,14 +130,6 @@
     goto err1;
   }
 
-  // check for copy permission
-  if (!doc->okToCopy()) {
-    error(errNotAllowed, -1,
-	  "Copying of text from this document is not allowed.");
-    exitCode = 3;
-    goto err1;
-  }
-
   // get page range
   if (firstPage < 1) {
     firstPage = 1;
--- a/xpdf/pdftotext.cc
+++ b/xpdf/pdftotext.cc
@@ -188,14 +188,6 @@
     goto err2;
   }
 
-  // check for copy permission
-  if (!doc->okToCopy()) {
-    error(errNotAllowed, -1,
-	  "Copying of text from this document is not allowed.");
-    exitCode = 3;
-    goto err2;
-  }
-
   // construct text file name
   if (argc == 3) {
     textFileName = new GString(argv[2]);
--- a/xpdf/XPDFCore.cc
+++ b/xpdf/XPDFCore.cc
@@ -384,12 +384,7 @@
 #ifndef NO_TEXT_SELECT
       if (selectULX != selectLRX &&
 	  selectULY != selectLRY) {
-	if (doc->okToCopy()) {
 	  copySelection();
-	} else {
-	  error(errNotAllowed, -1,
-		"Copying of text from this document is not allowed.");
-	}
       }
 #endif
     }
@@ -408,9 +403,6 @@
   int pg;
   double ulx, uly, lrx, lry;
 
-  if (!doc->okToCopy()) {
-    return;
-  }
   if (getSelection(&pg, &ulx, &uly, &lrx, &lry)) {
     //~ for multithreading: need a mutex here
     if (currentSelection) {
