--- openboardview-7.3/src/glad/utility/download.sh.orig	2018-10-31 18:28:09.700000000 -0000
+++ openboardview-7.3/src/glad/utility/download.sh	2018-10-31 18:38:22.136666653 -0000
@@ -3,16 +3,16 @@
 set -e
 
 rm -f egl.xml
-wget -O egl.xml https://cvs.khronos.org/svn/repos/ogl/trunk/doc/registry/public/api/egl.xml
+wget -O egl.xml https://raw.githubusercontent.com/capnm/Khronos_XML_API/master/egl.xml
 
 rm -f gl.xml
-wget -O gl.xml https://cvs.khronos.org/svn/repos/ogl/trunk/doc/registry/public/api/gl.xml
+wget -O gl.xml https://raw.githubusercontent.com/capnm/Khronos_XML_API/master/gl.xml
 
 rm -f glx.xml
-wget -O glx.xml https://cvs.khronos.org/svn/repos/ogl/trunk/doc/registry/public/api/glx.xml
+wget -O glx.xml https://raw.githubusercontent.com/capnm/Khronos_XML_API/master/glx.xml
 
 rm -f wgl.xml
-wget -O wgl.xml https://cvs.khronos.org/svn/repos/ogl/trunk/doc/registry/public/api/wgl.xml
+wget -O wgl.xml https://raw.githubusercontent.com/capnm/Khronos_XML_API/master/wgl.xml
 
 rm -f khrplatform.h
-wget -O khrplatform.h https://www.khronos.org/registry/egl/api/KHR/khrplatform.h
\ No newline at end of file
+wget -O khrplatform.h https://www.khronos.org/registry/egl/api/KHR/khrplatform.h
