--- a/Makefile.am	2015-01-30 14:24:07.000000000 -0500
+++ b/Makefile.am	2015-03-27 12:30:07.113123642 -0400
@@ -22,9 +22,6 @@
 MMCOREPY_DIR = MMCorePy_wrap
 endif
 
-if BUILD_SECRETDEVICEADAPTERS
-SECRETDEVICEADAPTERS = SecretDeviceAdapters
-endif
 
 
 if BUILD_JAVA_APP
@@ -126,7 +123,6 @@
 	$(MMCOREPY_DIR) \
 	$(JAVA_APP_DIRS) \
 	DeviceAdapters \
-	$(SECRETDEVICEADAPTERS) \
 	$(SYSTEMTEST_DIR)
 	bindist
 
--- a/configure.ac	2015-01-30 14:24:07.000000000 -0500
+++ b/configure.ac	2015-03-27 12:30:19.193123863 -0400
@@ -370,16 +370,8 @@
 ## Subdirectory configuration
 ##
 
-AC_MSG_CHECKING([for proprietary device adapter source])
-AS_IF([test -f "$srcdir/SecretDeviceAdapters/configure"],
-   [build_secretdeviceadapters=yes], [build_secretdeviceadapters=no])
-AM_CONDITIONAL([BUILD_SECRETDEVICEADAPTERS],
-   [test "x$build_secretdeviceadapters" = xyes])
-
+build_secretdeviceadapters=no
 AC_CONFIG_SUBDIRS([DeviceAdapters])
-AS_IF([test "x$build_secretdeviceadapters" = xyes], [
-   AC_CONFIG_SUBDIRS([SecretDeviceAdapters])
-])
 AC_CONFIG_SUBDIRS([DeviceKit])
 
 
