--- a/ThirdParty/GL/glh/glh_genext.h
+++ b/ThirdParty/GL/glh/glh_genext.h
@@ -26,6 +26,16 @@
 # define GLH_INITIALIZER
 #endif
 
+#ifndef GLX_NV_vertex_array_range
+#define GLX_NV_vertex_array_range 1
+#ifdef GLX_GLXEXT_PROTOTYPES
+extern void *glXAllocateMemoryNV (GLsizei, GLfloat, GLfloat, GLfloat);
+extern void glXFreeMemoryNV (void *);
+#endif /* GLX_GLXEXT_PROTOTYPES */
+typedef void * ( * PFNGLXALLOCATEMEMORYNVPROC) (GLsizei size, GLfloat readfreq, GLfloat writefreq, GLfloat priority);
+typedef void ( * PFNGLXFREEMEMORYNVPROC) (void *pointer);
+#endif
+
 #define GLH__PREPROCESSOR_GYMNASTICS2(a,b) a##b
 #define GLH__PREPROCESSOR_GYMNASTICS(a,b) GLH__PREPROCESSOR_GYMNASTICS2(a,b)
 
--- a/ThirdParty/PSCommon/XnLib/ThirdParty/GL/glh/glh_genext.h
+++ b/ThirdParty/PSCommon/XnLib/ThirdParty/GL/glh/glh_genext.h
@@ -26,6 +26,16 @@
 # define GLH_INITIALIZER
 #endif
 
+#ifndef GLX_NV_vertex_array_range
+#define GLX_NV_vertex_array_range 1
+#ifdef GLX_GLXEXT_PROTOTYPES
+extern void *glXAllocateMemoryNV (GLsizei, GLfloat, GLfloat, GLfloat);
+extern void glXFreeMemoryNV (void *);
+#endif /* GLX_GLXEXT_PROTOTYPES */
+typedef void * ( * PFNGLXALLOCATEMEMORYNVPROC) (GLsizei size, GLfloat readfreq, GLfloat writefreq, GLfloat priority);
+typedef void ( * PFNGLXFREEMEMORYNVPROC) (void *pointer);
+#endif
+
 #define GLH__PREPROCESSOR_GYMNASTICS2(a,b) a##b
 #define GLH__PREPROCESSOR_GYMNASTICS(a,b) GLH__PREPROCESSOR_GYMNASTICS2(a,b)
 
