--- a/cmdModule.c
+++ b/cmdModule.c
@@ -640,8 +640,9 @@
             case TCL_OK:	gotPartial = 0;
 			        continue;	/** while **/
 	    
-            case TCL_ERROR:	interp->errorLine = ((linenum-1)-gotPartial) +
-				    interp->errorLine;
+            case TCL_ERROR:	Tcl_SetErrorLine(interp,
+			((linenum-1)-gotPartial) +
+			Tcl_GetErrorLine(interp));
 	    			/* FALLTHROUGH */
 
             case TCL_LEVEL0_RETURN:
