--- libmpdemux/demux_ts.c.org	2009-03-16 23:31:49.000000000 -0400
+++ libmpdemux/demux_ts.c	2009-03-16 23:33:26.000000000 -0400
@@ -1344,7 +1344,7 @@
 
 static int pes_parse2(unsigned char *buf, uint16_t packet_len, ES_stream_t *es, int32_t type_from_pmt, pmt_t *pmt, int pid)
 {
-	unsigned char  *p;
+	unsigned char  *p, *o;
 	uint32_t       header_len;
 	int64_t        pts;
 	uint32_t       stream_id;
@@ -1359,7 +1359,7 @@
 		return 0;
 	}
 
-	p = buf;
+	o = p = buf;
 	pkt_len = packet_len;
 
 
@@ -1516,11 +1516,24 @@
 		if(type_from_pmt != UNKNOWN)
 		    es->type    = type_from_pmt;
 		else
-		    es->type    = VIDEO_MPEG2;
+		{
+			int i=0;
+			es->type    = VIDEO_MPEG2;
+
+			for (i=0; i < TS_PACKET_SIZE; i++)
+				if (o[i] == 0x00 && o[i+1] == 0x00 && o[i+2] == 0x01 && o[i+3] == 0x09)
+				{
+					es->type    = VIDEO_H264;
+					break;
+				}
+		}
 		if(es->payload_size)
 			es->payload_size -= packet_len;
 
-		mp_msg(MSGT_DEMUX, MSGL_DBG2, "pes_parse2: M2V size %d\n", es->size);
+		if (es->type == VIDEO_MPEG2)
+			mp_msg(MSGT_DEMUX, MSGL_DBG2, "pes_parse2: M2V size %d\n", es->size);
+		if (es->type == VIDEO_H264)
+			mp_msg(MSGT_DEMUX, MSGL_DBG2, "pes_parse2: H264V size %d\n", es->size);
 		return 1;
 	}
 	else if ((stream_id == 0xfa))
--- libmpdemux/demux_ts.h.org	2010-07-12 00:39:27.000000000 -0400
+++ libmpdemux/demux_ts.h	2010-07-12 00:39:35.000000000 -0400
@@ -21,7 +21,7 @@
 
 #include <stdint.h>
 
-#define TS_MAX_PROBE_SIZE 2000000
+#define TS_MAX_PROBE_SIZE 9000000
 
 int mp_a52_framesize(uint8_t *buf, int *srate);
 
