--- a/lib/libxdp/libxdp.c
+++ b/lib/libxdp/libxdp.c
@@ -1007,7 +1007,10 @@
 	}
 
 	if (info.btf_id && !xdp_prog->btf) {
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wdeprecated-declarations"
 		err = btf__get_from_id(info.btf_id, &btf);
+#pragma GCC diagnostic pop
 		if (err) {
 			pr_warn("Couldn't get BTF for ID %ul\n", info.btf_id);
 			goto err;
@@ -1705,7 +1708,10 @@
 	arrays = (1UL << BPF_PROG_INFO_MAP_IDS);
 
 	if (prog_fd > 0) {
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wdeprecated-declarations"
 		info_linear = bpf_program__get_prog_info_linear(prog_fd, arrays);
+#pragma GCC diagnostic pop
 		if (IS_ERR_OR_NULL(info_linear)) {
 			pr_warn("couldn't get program info for fd: %d", prog_fd);
 			return -EINVAL;
@@ -1718,7 +1724,10 @@
 			goto legacy;
 		}
 
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wdeprecated-declarations"
 		err = btf__get_from_id(info->btf_id, &btf);
+#pragma GCC diagnostic pop
 		if (err) {
 			pr_warn("Couldn't get BTF for ID %ul\n", info->btf_id);
 			goto out;
@@ -2039,7 +2048,10 @@
 		pr_warn("The target program doesn't have BTF\n");
 		return -EINVAL;
 	}
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wdeprecated-declarations"
 	if (btf__get_from_id(info.btf_id, &btf)) {
+#pragma GCC diagnostic pop
 		pr_warn("Failed to get BTF of the program\n");
 		return -EINVAL;
 	}
