commit 81c036acf45a8e29490048d8bfca7c211df07105
Author: xdeller <xdeller@z820.int.xdel.org>
Date:   Sat Jul 27 19:00:11 2019 +0300

    Adapt microvm patch series to current 4.1.0-rc2

diff --git a/hw/i386/microvm.c b/hw/i386/microvm.c
index b3b367add1..dac173f752 100644
--- a/hw/i386/microvm.c
+++ b/hw/i386/microvm.c
@@ -164,7 +164,7 @@ static void microvm_cpus_init(const char *typename, Error **errp)
 {
     int i;
 
-    for (i = 0; i < smp_cpus; i++) {
+    for (i = 0; i < current_machine->smp.cpus; i++) {
         Object *cpu = NULL;
         Error *local_err = NULL;
 
@@ -439,7 +439,7 @@ static void microvm_mptable_setup(MicrovmMachineState *mms)
     char *mptable;
     int size;
 
-    mptable = mptable_generate(smp_cpus, EBDA_START, &size);
+    mptable = mptable_generate(current_machine->smp.cpus, EBDA_START, &size);
     address_space_write(&address_space_memory,
                         EBDA_START, MEMTXATTRS_UNSPECIFIED,
                         (uint8_t *) mptable, size);
@@ -460,9 +460,8 @@ static void microvm_machine_set_legacy(Object *obj, bool value, Error **errp)
     mms->legacy = value;
 }
 
-static void microvm_machine_reset(void)
+static void microvm_machine_reset(MachineState *machine)
 {
-    MachineState *machine = MACHINE(qdev_get_machine());
     MicrovmMachineState *mms = MICROVM_MACHINE(machine);
     CPUState *cs;
     X86CPU *cpu;
