diff -Nrup a/pgadmin/include/utils/pgDefs.h b/pgadmin/include/utils/pgDefs.h
--- a/pgadmin/include/utils/pgDefs.h	2016-01-07 05:47:32.000000000 -0700
+++ b/pgadmin/include/utils/pgDefs.h	2019-09-24 13:31:11.389591310 -0600
@@ -18,9 +18,9 @@
 
 #define PGOID_CLASS_PG_AUTHID               1260L
 
-#define PGOID_TYPE_SERIAL                   -42L
-#define PGOID_TYPE_SERIAL8                  -43L
-#define PGOID_TYPE_SERIAL2                  -44L
+#define PGOID_TYPE_SERIAL                   -42uL
+#define PGOID_TYPE_SERIAL8                  -43uL
+#define PGOID_TYPE_SERIAL2                  -44uL
 #define PGOID_TYPE_BOOL                     16L
 #define PGOID_TYPE_BYTEA                    17L
 #define PGOID_TYPE_CHAR                     18L
